### DISCUSSION TITLE
mac 上使用  backend = transformers  报错

### DISCUSSION THREAD

**Discussion by kenneth-bro** [magicpdf/Magic-PDF#2982] [Q&A]
### 🔎 Search before asking | 提交之前请先搜索

- [x] I have searched the MinerU [Readme](https://github.com/opendatalab/MinerU) and found no similar bug report.
- [x] I have searched the MinerU [Issues](https://github.com/opendatalab/MinerU/issues) and found no similar bug report.
- [x] I have searched the MinerU [Discussions](https://github.com/opendatalab/MinerU/discussions) and found no similar bug report.

### 🤖 Consult the online AI assistant for assistance | 在线 AI 助手咨询

- [x] I have consulted the [online AI assistant](https://deepwiki.com/opendatalab/MinerU) but was unable to obtain a solution to the issue.

### Description of the bug | 错误描述

<img width="1096" height="154" alt="Image" src="https://github.com/user-attachments/assets/547d625a-c242-45b1-8659-71604388f596" />

2025-07-11 09:52:44.152 | INFO     | mineru.backend.vlm.predictor:get_predictor:110 - get_predictor cost: 1.64s
Predict:   0%|                                                                 | 0/13 [00:00<?, ?it/s]The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.

就这样不动了

### How to reproduce the bug | 如何复现

使用fastapi的方式启动服务并请求。
curl -X 'POST' \
  'http://192.168.48.10:8999/file_parse' \
  -H 'accept: application/json' \
  -H 'Content-Type: multipart/form-data' \
  -F 'return_middle_json=false' \
  -F 'return_model_output=false' \
  -F 'return_md=true' \
  -F 'return_images=false' \
  -F 'end_page_id=99999' \
  -F 'parse_method=auto' \
  -F 'start_page_id=0' \
  -F 'lang_list=en' \
  -F 'output_dir=./output' \
  -F 'server_url=string' \
  -F 'return_content_list=false' \
  -F 'backend=transformers' \
  -F 'table_enable=true' \
  -F 'files=@prepped.pdf;type=application/pdf' \
  -F 'formula_enable=true'

### Operating System Mode | 操作系统类型

MacOS

### Operating System Version| 操作系统版本

15.1

### Python version | Python 版本

3.12

### Software version | 软件版本 (mineru --version)

2.0.x

### Device mode | 设备模式

mps

**Comment by dosubot** [comment]
你的问题是因为 MinerU 的 transformers 后端在推理时没有显式传递 attention_mask，而 transformers 库在 pad_token_id 和 eos_token_id 相同的情况下，无法自动推断 attention_mask，这会导致模型行为异常甚至卡死，尤其是在 MacOS + MPS（Apple GPU）环境下更容易复现这个问题。

建议的解决办法：

1. 修改 mineru/backend/vlm/hf_predictor.py 里的 HuggingfacePredictor，将 tokenizer 返回的 attention_mask 一起传递给 model.generate。例如：

```python
inputs = self.tokenizer(prompt, return_tensors="pt")
input_ids = inputs.input_ids.to(device=self.model.device)
attention_mask = inputs.attention_mask.to(device=self.model.device)
output_ids = self.model.generate(
    input_ids,
    attention_mask=attention_mask,
    images=image_tensor,
    image_sizes=image_sizes,
    use_cache=True,
    **generate_kwargs,
    **kwargs,
)
```

2. 检查你用的 tokenizer 的 pad_token_id 和 eos_token_id，尽量不要让它们相同。如果模型支持，设置不同的 pad_token_id 和 eos_token_id。

3. 如果只是 Mac 上 MPS 卡死，可以尝试切换 device_mode 为 cpu，看是否能正常运行，帮助定位是 MPS 相关 bug 还是 attention_mask 问题。

4. 目前 MinerU 代码中 attention_mask 处理有 FIXME 注释，说明这块实现还不完善，建议优先用上述 workaround 规避 [代码参考](https://github.com/opendatalab/MinerU/blob/5f1175fadb8587a9d8a6a007b64b1228dd7348b8/mineru/backend/vlm/hf_predictor.py)。

这个 warning 不是简单的提示，确实会导致推理异常甚至死锁，尤其是 Mac/MPS 环境。建议尽快修正 attention_mask 的传递。



*To reply, just mention [@dosu](https://go.dosu.dev/dosubot).*

---
<sup>How did I do? [Good](https://app.dosu.dev/response-feedback/a87ab983-8f9f-49e4-bd9a-19b43bcea9bd?feedback_type=great_response) | [Irrelevant](https://app.dosu.dev/response-feedback/a87ab983-8f9f-49e4-bd9a-19b43bcea9bd?feedback_type=irrelevant_answer) | [Incorrect](https://app.dosu.dev/response-feedback/a87ab983-8f9f-49e4-bd9a-19b43bcea9bd?feedback_type=incorrect_sources) | [Verbose](https://app.dosu.dev/response-feedback/a87ab983-8f9f-49e4-bd9a-19b43bcea9bd?feedback_type=too_verbose) | [Hallucination](https://app.dosu.dev/response-feedback/a87ab983-8f9f-49e4-bd9a-19b43bcea9bd?feedback_type=hallucination) | [Report 🐛](https://app.dosu.dev/response-feedback/a87ab983-8f9f-49e4-bd9a-19b43bcea9bd?feedback_type=bug_report) | [Other](https://app.dosu.dev/response-feedback/a87ab983-8f9f-49e4-bd9a-19b43bcea9bd?feedback_type=other)</sup>&nbsp;&nbsp;[![Join Discord](https://img.shields.io/badge/join-5865F2?logo=discord&logoColor=white&label=)](https://go.dosu.dev/discord-bot)&nbsp;[![Share on X](https://img.shields.io/badge/X-share-black)](https://twitter.com/intent/tweet?text=%40dosu_ai%20helped%20me%20solve%20this%20issue!&url=https%3A//github.com/opendatalab/MinerU/issues/2977)

**Comment by dosubot** [comment]
你遇到的 warning（attention mask 未设置且 pad token 和 eos token 相同）是 HuggingFace Transformers 在输入没有明确 attention_mask 时的典型提示。如果 pad_token 和 eos_token 相同，模型无法自动推断哪些 token 需要被 mask，这会导致推理异常，尤其是在 Mac MPS（Apple Metal）环境下，PyTorch/Transformers 的兼容性本身就不如 CUDA，容易出现卡死或无响应的情况。

根据 MinerU 的实现，目前 HuggingfacePredictor 并没有强制生成和传递 attention_mask，而是依赖 transformers 自动推断，这在你的场景下会触发上述 warning 并可能导致挂起[代码参考](https://github.com/opendatalab/MinerU/blob/5f1175fadb8587a9d8a6a007b64b1228dd7348b8/mineru/backend/vlm/predictor.py)。

建议尝试以下方法规避问题：
1. 检查你用的模型和 tokenizer，确保 pad_token_id 和 eos_token_id 不要设置成同一个值（可以在加载 tokenizer 后手动设置）。
2. 修改 MinerU 相关代码，在调用 transformers 生成/推理时，显式传递 attention_mask（可通过 tokenizer.encode_plus 或 tokenizer.batch_encode_plus 生成）。
3. 如果短期内无法修改代码，可以尝试切换 device 到 cpu，看是否能正常推理（MPS 兼容性问题较多）。
4. 检查 transformers 和 torch 版本，升级到最新版有时能解决部分 MPS 相关 bug。

如果需要具体代码修改建议，可以补充你用的模型名和 tokenizer 配置。



*To reply, just mention [@dosu](https://go.dosu.dev/dosubot?utm_source=github&utm_medium=bot-comment&utm_campaign=github-comment-footer-20260415&utm_content=reply-with-mention&utm_term=opendatalab%2FMinerU).*

---

Docs are dead. Just use [Dosu](https://dosu.dev?utm_source=github&utm_medium=bot-comment&utm_campaign=github-comment-footer-20260415&utm_content=static-docs-tagline&utm_term=opendatalab%2FMinerU).

[![Leave Feedback](https://img.shields.io/badge/Leave%20Feedback-555555?style=flat)](https://app.dosu.dev/response-feedback?utm_source=github&utm_medium=bot-comment&utm_campaign=github-comment-footer-20260415&utm_content=static-docs-feedback&utm_term=opendatalab%2FMinerU&message_id=4a0c48ff-e7e4-446e-b5d2-c0452fc7ce99) [![Ask Dosu about MinerU](https://img.shields.io/badge/Ask%20Dosu%20about%20MinerU-2f7b3f?style=flat&logo=data%3Aimage%2Fsvg%2Bxml%3Bbase64%2CPHN2ZyB3aWR0aD0iODYiIGhlaWdodD0iODkiIHZpZXdCb3g9IjAgMCA4NiA4OSIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48cGF0aCBkPSJNNS4yOTIzNiAxMi43OTI4TDE3Ljc1OTMgNi42ODE4OFY3Mi41NjY3TDUuMjkyMzYgODQuMDYxOFYxMi43OTI4WiIgZmlsbD0iI0I0QkI5MSIvPjxwYXRoIGQ9Ik0xOC4yNTc1IDczLjExOTZMNTkuMTMyOSA3Mi43NDhMNTEuNzAxMSA4Mi40MDk1TDI5LjAzMzggODYuMjkxTDYuMjM5NjIgODUuMTU1NEwxOC4yNTc1IDczLjExOTZaIiBmaWxsPSIjNzc4NTYxIi8%2BPHBhdGggZD0iTTE3LjQ5MTYgMy43MzYzM0wzLjU4NTU3IDEyLjcwOTlWODMuNTc5MkMzLjU4NTU3IDg0Ljc1NDIgNC45ODU2MyA4NS4zNjUyIDUuODQ3MDUgODQuNTY2TDE5LjYyOTYgNzEuNzgwMSIgc3Ryb2tlPSJibGFjayIgc3Ryb2tlLXdpZHRoPSI2LjQyODQ0IiBzdHJva2UtbGluZWNhcD0icm91bmQiLz48bWFzayBpZD0iZG9zdS1kLWN1dG91dCIgZmlsbD0id2hpdGUiPjxwYXRoIGZpbGwtcnVsZT0iZXZlbm9kZCIgY2xpcC1ydWxlPSJldmVub2RkIiBkPSJNNDAuNzA0IDAuNTE4MDY2SDE3LjA0MzlWNzYuMjIyMUg0MC43MDRINDIuNTgwNUg0Ny44MDEzQzY4LjcwNjQgNzYuMjIyMSA4NS42NTMzIDU5LjI3NTIgODUuNjUzMyAzOC4zNzAxQzg1LjY1MzMgMTcuNDY1IDY4LjcwNjMgMC41MTgwNjYgNDcuODAxMyAwLjUxODA2Nkg0Mi41ODA1SDQwLjcwNFoiLz48L21hc2s%2BPHBhdGggZmlsbC1ydWxlPSJldmVub2RkIiBjbGlwLXJ1bGU9ImV2ZW5vZGQiIGQ9Ik00MC43MDQgMC41MTgwNjZIMTcuMDQzOVY3Ni4yMjIxSDQwLjcwNEg0Mi41ODA1SDQ3LjgwMTNDNjguNzA2NCA3Ni4yMjIxIDg1LjY1MzMgNTkuMjc1MiA4NS42NTMzIDM4LjM3MDFDODUuNjUzMyAxNy40NjUgNjguNzA2MyAwLjUxODA2NiA0Ny44MDEzIDAuNTE4MDY2SDQyLjU4MDVINDAuNzA0WiIgZmlsbD0iI0YzRjZGMSIvPjxwYXRoIGQ9Ik0xNy4wNDM5IDAuNTE4MDY2Vi02LjU3OTE5SDkuOTQ2NjlWMC41MTgwNjZIMTcuMDQzOVpNMTcuMDQzOSA3Ni4yMjIxSDkuOTQ2NjlWODMuMzE5NEgxNy4wNDM5Vjc2LjIyMjFaTTE3LjA0MzkgNy42MTUzMkg0MC43MDRWLTYuNTc5MTlIMTcuMDQzOVY3LjYxNTMyWk0yNC4xNDEyIDc2LjIyMjFWMC41MTgwNjZIOS45NDY2OVY3Ni4yMjIxSDI0LjE0MTJaTTQwLjcwNCA2OS4xMjQ5SDE3LjA0MzlWODMuMzE5NEg0MC43MDRWNjkuMTI0OVpNNDIuNTgwNSA2OS4xMjQ5SDQwLjcwNFY4My4zMTk0SDQyLjU4MDVWNjkuMTI0OVpNNDcuODAxMyA2OS4xMjQ5SDQyLjU4MDVWODMuMzE5NEg0Ny44MDEzVjY5LjEyNDlaTTc4LjU1NiAzOC4zNzAxQzc4LjU1NiA1NS4zNTU1IDY0Ljc4NjcgNjkuMTI0OSA0Ny44MDEzIDY5LjEyNDlWODMuMzE5NEM3Mi42MjYxIDgzLjMxOTQgOTIuNzUwNSA2My4xOTQ5IDkyLjc1MDUgMzguMzcwMUg3OC41NTZaTTQ3LjgwMTMgNy42MTUzMkM2NC43ODY2IDcuNjE1MzIgNzguNTU2IDIxLjM4NDcgNzguNTU2IDM4LjM3MDFIOTIuNzUwNUM5Mi43NTA1IDEzLjU0NTMgNzIuNjI2IC02LjU3OTE5IDQ3LjgwMTMgLTYuNTc5MTlWNy42MTUzMlpNNDIuNTgwNSA3LjYxNTMySDQ3LjgwMTNWLTYuNTc5MTlINDIuNTgwNVY3LjYxNTMyWk00MC43MDQgNy42MTUzMkg0Mi41ODA1Vi02LjU3OTE5SDQwLjcwNFY3LjYxNTMyWiIgZmlsbD0iYmxhY2siIG1hc2s9InVybCgjZG9zdS1kLWN1dG91dCkiLz48cGF0aCBkPSJNNjguOTIxNSAzNi4wMTM1QzY4LjkyMTUgMzYuMDEzNSA2NS43MzY5IDQ5LjQ3MzggNTEuNDIzMSA0OS40NzM4QzM3LjEwOTMgNDkuNDczOCAzMi41Nzg3IDM3LjM1OTYgMzIuNTc4NyAzNi4wMTM1IiBzdHJva2U9ImJsYWNrIiBzdHJva2Utd2lkdGg9IjcuNjkxNjEiIHN0cm9rZS1saW5lY2FwPSJyb3VuZCIgc3Ryb2tlLWxpbmVqb2luPSJyb3VuZCIvPjxwYXRoIGQ9Ik0wLjM0ODYzMyA4NS40OTQ2QzAuMzQ4NjMzIDg1LjQ5NDYgMjkuNDg1NiA4NS44MzA5IDM0LjgwOSA4NS42OThDNDQuODMzNyA4NS40NDc3IDUxLjI4NzIgODQuNDAyIDU3LjUyNjkgNzguOTcyNEM2Mi44MTI5IDc0LjM3MjcgNzUuMTM0MiA1OS42ODM2IDc1LjEzNDIgNTkuNjgzNiIgc3Ryb2tlPSJibGFjayIgc3Ryb2tlLXdpZHRoPSI2LjE2NDgyIi8%2BPC9zdmc%2B)](https://github.dosu.com/opendatalab/MinerU?utm_source=github&utm_medium=bot-comment&utm_campaign=github-comment-footer-20260415&utm_content=static-docs-ask-repo&utm_term=opendatalab%2FMinerU) [![Share Dosu with your team](https://img.shields.io/badge/Share%20Dosu%20with%20your%20team-1f6feb?style=flat)](https://app.dosu.dev/signup?referrer=openSource&source=github-footer&utm_source=github&utm_medium=bot-comment&utm_campaign=github-comment-footer-20260415&utm_content=static-docs-share-team&utm_term=opendatalab%2FMinerU)

**Comment by myhloli** [comment]
readme里有写，vlm需要n卡运行

**Reply by kenneth-bro** [reply]
@myhloli Linux和macOS系统安装后自动支持cuda/mps加速  这个是pipeline的？

**Reply by myhloli** [reply]
对，是pipeline的
